### PR TITLE
Unstably allow assume intrinsic in const contexts

### LINKED ISSUE #49572: The NonZero types don't tell LLVM that they're non-zero on get
Repro: https://play.rust-lang.org/?gist=0981601dda9c5800e353e31b22682bb5&version=nightly
```rust
pub fn foo(x: std::num::NonZeroU32) -> bool {
    x.get() != 0
}
```
Actual:
```llvm
  %0 = icmp ne i32 %x, 0
  ret i1 %0
```
Expected:
```llvm
  ret i1 true
```
cc https://github.com/rust-lang/rust/issues/49137

### PR DESCRIPTION
**Pull request by tesuji**
Not sure much about this usage because there are concerns 
about [blocking  optimization][1] and [slowing down LLVM][2] when using `assme` intrinsic
in inline functions.
But since Oli suggested in https://github.com/rust-lang/rust/issues/76960#issuecomment-695772221,
here we are.

[1]: https://github.com/rust-lang/rust/pull/54995#issuecomment-429302709
[2]: https://github.com/rust-lang/rust/issues/49572#issuecomment-589615423

### CODE REVIEW

File: src/test/ui/consts/const-eval/const_assume.rs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+// Check that `const_assume` feature allows to `assume`
+// to be used in const contexts.
+
+#![feature(core_intrinsics, const_assume)]
+
+extern crate core;
+
+use core::intrinsics::assume;
+
+pub const fn foo(x: usize, y: usize) -> usize {
```
Comment (oli-obk) [inline src/test/ui/consts/const-eval/const_assume.rs:12]: :laughing: please make this function unsafe. I know it's not called, but still

File: compiler/rustc_mir/src/interpret/intrinsics.rs [hunk 2]
```diff
@@ -435,6 +435,12 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 // These just return their argument
                 self.copy_op(args[0], dest)?;
             }
+            sym::assume => {
+                let cond = self.read_scalar(args[0])?.check_init()?.to_bool()?;
+                if !cond {
+                    throw_ub_format!("`assume` intrinsic called with `false`");
+                }
+            }
```
Comment (RalfJung) [inline compiler/rustc_mir/src/interpret/intrinsics.rs:443]: Please remember to also remove [the Miri implementation of this](https://github.com/rust-lang/miri/blob/1f6ce43721bb13afeda2b91dea2db80b89cea37b/src/shims/intrinsics.rs#L481) once the PR lands.

File: src/test/ui/consts/const-eval/const_assume.rs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+#![feature(core_intrinsics, const_assume)]
+
+extern crate core;
+
+use core::intrinsics::assume;
+
+pub const unsafe fn foo(x: usize, y: usize) -> usize {
+    assume(y != 0);
+    x / y
+}
```
Comment (RalfJung) [inline src/test/ui/consts/const-eval/const_assume.rs:15]: Isn't this the kind of test that should be made a unit test instead (https://github.com/rust-lang/rust/issues/76268)?

File: compiler/rustc_mir/src/interpret/intrinsics.rs (hunk 2 above)
Comment (lzutao) [reply compiler/rustc_mir/src/interpret/intrinsics.rs:443]: yeah, I'm preparing PR for that.

File: src/test/ui/consts/const-eval/const_assume.rs (hunk 3 above)
Comment (lzutao) [reply src/test/ui/consts/const-eval/const_assume.rs:15]: I only saw `const_slice_from_raw_parts` and `const_raw_ptr_deref` in `library/core/tests/lib.rs`.
Don't really sure about this.

File: library/core/src/num/nonzero.rs [hunk 4]
```diff
@@ -81,7 +82,12 @@ assert_eq!(size_of::<Option<core::num::", stringify!($Ty), ">>(), size_of::<", s
                 #[$stability]
                 #[inline]
                 #[rustc_const_stable(feature = "nonzero", since = "1.34.0")]
+                #[allow_internal_unstable(const_assume)]
                 pub const fn get(self) -> $Int {
+                    // SAFETY: inner value of `NonZero*` type should never be zero.
```
Comment (oli-obk) [inline library/core/src/num/nonzero.rs:87]:
```suggestion
                    // SAFETY: inner value of `NonZero*` type can never be zero.
```

**Comment by lzutao** [comment]
Not sure the block optimization concerns about adding assume to `NonZero*::get` still hold: https://github.com/rust-lang/rust/pull/54995#issuecomment-429302709

**Comment by lzutao** [comment]
Because the above concern isn't resolved, I will move the `assume` part in `NonZero*::get` to another PR.

**Comment by oli-obk** [comment]
Yea they probably do :( I don't think there's any point in summoning nagisa or hanna, I don't think anything changed over the last two years that woudl affect this.

File: src/test/ui/consts/const-eval/const_assume.rs (hunk 3 above)
Comment (matklad) [reply src/test/ui/consts/const-eval/const_assume.rs:15]: Yeah, I think this should be a unit-test.

Comment (lzutao) [reply src/test/ui/consts/const-eval/const_assume.rs]: It's kind of weird since most intrinsic not tested in `library/core/tests`.
Any, I moved it.

**Pull request merged**